### DISCUSSION TITLE
Processor [:vips] not reading base64

### DISCUSSION THREAD

**Discussion by jessebarnes** [libvips/ruby-vips#280] [Q&A]
Have a Posts.rb with has_attached_file:
```
 has_attached_file :gif, 
    styles: {
        original: {},
        large: '512x512>', format: 'gif',
        square: '256x256>',  format: 'gif',
        thumbnail: '70x70>', animated: false, format: 'gif' 
      },
      :convert_options => {
        :cmyk_original => lambda { |file| [{
            cmd: "copy",
            args: ["n: -1"]
        }].to_json }
    },
    processors: [:vips]

    validates_attachment :gif, content_type: { content_type: 'image/gif' }, if: Proc.new { |a| a.gif? }
    validates_with AttachmentSizeValidator, attributes: :gif,
    less_than: 9.megabytes,
    message: 'Gif exceeds 9mb upload size limit'

```
Postman API Body:
``` 
"content": "Hi Mom",
"gif": "data:image/gif;base64,R0lGODlhAALXAPcAAAECBAsEAwAGC.."
```
With Processor [:vips] we received : 
```
{
    "error": true,
    "message": "Gif Paperclip::Errors::NotIdentifiedByImageMagickError, Gif has contents that are not what they are reported to be"
}
```

When the processor is taken out, we do not have any issues.  Not sure what is happening... any suggestions?  Is my model correct?  Thanks ~

**Comment by jcupitt** [comment]
Hi @jessebarnes, this looks like a ImageProcessing (I think?) error. I'd open an issue there.

**Reply by jessebarnes** [reply]
[Hi] @jcupitt !

I worked around it(decode method before post process)! Thank you for the help, but my gif saves just the first image of my gif.. Is there something wrong with my options?

```
 has_attached_file :gif, 
    styles: {
        original: {},
        large: '512x512>', format: 'gif',
        square: '256x256>',  format: 'gif',
        thumbnail: '70x70>', animated: false, format: 'gif' 
      },
      :convert_options => {
        :cmyk_original => lambda { |file| [{
            cmd: "copy",
            args: ["n: -1"]
        }].to_json }
    },
    processors: [:vips]
```

Or should it be

```
styles: {
        original: {},
        large: { geometry: '512x512>', 'animated' => true, 'animated.delay' => 4},
        square:   { geometry: '256x256>',  format: 'gif'},
        thumbnail: { geometry: '70x70>', format: 'gif' }
       },
    convert_options: {
      :cmyk_original => lambda { |file| [{
        cmd: "icc_transform",
        args: ["copy [n=-1]"],
        optional: {
            intent: :sequential
        },
    }].to_json }
    },  
    processors: [:vips]
```
Please help!~
<img width="1254" alt="Screen Shot 2021-02-24 at 1 47 22 PM" src="https://user-images.githubusercontent.com/69123787/109050306-3d00a300-76a7-11eb-898f-3b2edf14e4b5.png">

**Reply by jcupitt** [reply]
You need to set `n: -1` to get all frames (libvips just does a single frame by default), so I guess there's some problem with the way you're passing that argument.

But I don't know what the exact error is, I've never used the framework you're using.

**Reply by jessebarnes** [reply]
trying to pass in Post model w/convert_options on has_attached_file :gif in ruby on rails

when I pass in "copy" in cmd and "n:-1" as args, I get this error:
"unable to call copy: you supplied 2 arguments, but operation needs 1"

current code:
```
   convert_options: {
                          :large => lambda { |file| [{
                            cmd: "copy",
                            args: "[n: -1]"
                          }].to_json }
                   },
```
_error: unable to call copy: you supplied 2 arguments, but operation needs 1_

Thanks so much @jcupitt, am I passing this in incorrectly?

**Reply by jcupitt** [reply]
Sorry, no idea, you are not using ruby-vips, you are using the ImageProcessing gem (I think). You'd need to ask there.

**Reply by jessebarnes** [reply]
No, I am using Ruby-Vips. I am using your gem with Image Magick and kt-paper-clip. The above usage is from your documentation.  

How would you incorporate your gem with this argument in a model like the example usage shows in your readme?

Please, help~
 
Gemfile:
ruby-vips (2.0.17)
      ffi (~> 1.9)
 ...
paperclip-vips!
remote: https://github.com/tsu-social/paperclip-vips.git
...
here is the github link ->  https://github.com/realhub/paperclip-vips 

Is this the wrong usage? thank you so much

**Reply by jcupitt** [reply]
Yes, but you are not using ruby-vips directly, you're using it via some other API I don't know, perhaps the ImageProcessing gem, or maybe paperclip.

You need to ask the people who made the gem you are using.

**Reply by jcupitt** [reply]
Looks like you need to ask paperclip-vips, that's the gem you are using.

Though I don't know if it's the best solution -- in rails6, ActiveRecord uses ImageProcessing for image handling, and that has a vips backend. I think I would go that route instead.

Have a look at eg.:

https://bloggie.io/@kinopyo/upgrade-guide-active-storage-in-rails-6
